src/lib/y2partitioner/actions/clone_disk.rb

```diff
@@ -0,0 +1,115 @@
+# encoding: utf-8
+
+# Copyright (c) [2018] SUSE LLC
+#
+# All Rights Reserved.
+#
+# This program is free software; you can redistribute it and/or modify it
+# under the terms of version 2 of the GNU General Public License as published
+# by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
+# more details.
+#
+# You should have received a copy of the GNU General Public License along
+# with this program; if not, contact SUSE LLC.
+#
+# To contact SUSE LLC about this file by physical or electronic mail, you may
+# find current contact information at www.suse.com.
+
+require "yast"
+require "yast/i18n"
+require "y2partitioner/dialogs/disk_clone"
+require "y2partitioner/actions/controllers/disk_device"
+
+Yast.import "Popup"
+
+module Y2Partitioner
+  module Actions
+    # Action for cloning a disk
+    #
+    # To clone a disk means to copy the partition table and its partitions. Filesystems,
+    # encryptions, LVM or MD RAID over the partitions are not copied.
+    class CloneDisk
+      include Yast::I18n
+
+      # Constructor
+      #
+      # @param device [Y2Storage::Disk, Y2Storage::Dasd, Y2Storage::Multipath, Y2Storage::DmRaid]
+      def initialize(device)
+        textdomain "storage"
+
+        @controller = Controllers::DiskDevice.new(device)
+      end
+
+      # Checks whether it is possible to clone the device, and if so, the action is performed.
+      #
+      # @note An error popup is shown when the device cannot be cloned. In case the device
+      #   can be cloned, a dialog is presented to select over which disks to clone. Selected
+      #   disks are stored into the controller, see
+      #   {Controllers::DiskDevice#selected_devices_for_cloning}).
+      #
+      # @return [Symbol] :finish if the action is performed; :back or dialog result otherwise.
+      def run
+        return :back unless validate
+
+        result = Dialogs::DiskClone.run(controller)
+        return result if result != :ok
+
+        clone_disk
+        :finish
+      end
+
+    private
+
+      # @return [Controllers::DiskDevice]
+      attr_reader :controller
+
+      # Checks whether the clone action can be performed
+      #
+      # @note A disk can be cloned if it has a partition table and there are other
+      #   suitable disks where to clone it. In case the disk cannot be cloned, an
+      #   error message is shown.
+      #
+      # @return [Boolean] true if the clone action can be performed; false otherwise.
+      def validate
+        error = partition_table_error || suitable_devices_error
+        return true if error.nil?
```

Comment:
what about `return true unless error`?

Reply:
Well, I agree `unless error` is more readable, but in general I prefer to distinguish between nil and false cases. With `error.nil?` we explicitly express that there is no error. With `unless error` we are not sure at first glance if error is boolean or not.

Only a matter of taste.

Reply:
well, for me explicit nil? usually indicates that value can be `false` which is not true in this case.

Reply:
Different point of view. I use it to check that the object is pointing to "nothing".